### PR TITLE
KFP_ARGO_COMPILE_WORKFLOW_ONLY feature

### PR DESCRIPTION
**Pull request by talebzeghmi**
This allows CICD to compile a `Workflow` for an ephemeral run without creating lasting resources such as `ConfigMap`, `CronWorkflow` in the cluster, especially used for `train_and_evaluate` meant for training and evaluating on small data for quick validation.

### CODE REVIEW

File: metaflow/plugins/kfp/kfp.py
Comment (cloudw) [inline metaflow/plugins/kfp/kfp.py:411]: As we discussed before, moving the logic for when to create cron workflow to aip-doit-module or CICD code may be cleaner. 

Instead of disabling CronWorkflow on a variable, can we by default not create a CronWorkflow when running a Workflow?

File: metaflow/plugins/kfp/kfp.py
Comment (alexlatchford) [inline metaflow/plugins/kfp/kfp.py:311]: Looks like I missed earlier PRs that introduced this feature, just to make sure I'm understanding how we're using this feature, looks like we're using [Workflow-level synchronization](https://argoproj.github.io/argo-workflows/synchronization/) right? So what this means in practice is if this `Workflow` runs in any capacity (either ad-hoc, triggered by an event or from a `CronWorkflow`) it'd be bound to the `max_run_concurrency`?

Looking at the top of the file seems like we limit `max_run_concurrency` to `10` by default too?

File: metaflow/plugins/kfp/kfp.py
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:411]: We don't create a CronWorkflow when running a Workflow, see 

https://github.com/zillow/metaflow/blob/tz/AIP-7300-argo-cicd/metaflow/plugins/kfp/kfp.py#L356-L386

File: metaflow/plugins/kfp/kfp.py
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:311]: Yes! and the synchronization semaphore is shared between (WorkflowTemplate triggers,  CronWorkflow runs, and ad-hoc Workflow submissions)

**Review by alexlatchford — APPROVED** [review]
Nothing blocking from me, will defer to Yun though as my knowledge of this codebase is definitely aging 😅

File: metaflow/plugins/kfp/kfp.py
Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:411]: I'm still leaning towards not nesting the logic of creating CronWorkflow inside create_workflow_yaml_file. I tend to think of metaflow sdk as a lower level interface that contain minimal amount of assumption on how the functions would be called. Do you think line 416-423 can be moved to a separate function called `create_cron_workflow_yaml_file`?

**Comment by cloudw** [comment]
Though I have concern over where the cron workflow code is put, that doesn't seem to be inside the scope of this PR. Approving...

**Review by cloudw — APPROVED** [review]

File: metaflow/plugins/kfp/kfp.py
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:411]: The idea is to:
1) Follow Metaflow Argo Plugin
2) Behaviors about a Flow can all be defined as step and flow decorators.
3) The Flow can be compiled into a single yaml representing the behavior declared in python Flow file.

That is why all related behavior is onefile:

1. The argo workflow
2. The configmap containing the max concurrent runs synchronization semaphore value.
3. The argo CronWorkflow

Comment (cloudw) [reply metaflow/plugins/kfp/kfp.py:411]: I don't think I want to block this PR further as this is actually a more minor issue.

Though I can try to explain my preference in more clarity. My ask is mostly a small refactoring for code structure.

I think low level interface (`kfp.py`) should do exactly one thing per function. For example, `netflix/metaflow` has [register_workflow_template](https://github.com/Netflix/metaflow/blob/ec8fd6c4a161b13a0f3765efb1e7fea92cdc2baa/metaflow/plugins/argo/argo_client.py#L55) and [schedule_workflow_template](https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/argo/argo_client.py#L270) separately.

At layers where user scenario is clear (`kfp_cli.py` or `ci-cd-template` corresponding to notebook or CI pipeline deployment), we can then have functions combining multiple lower level interface, such as deploying (WorkflowTemplate) and scheduling (CronWorkflow) the workflow in one call. For example, `netflix/metaflow` combines flow deploy and trigger call in one [create](https://github.com/Netflix/metaflow/blob/ec8fd6c4a161b13a0f3765efb1e7fea92cdc2baa/metaflow/plugins/argo/argo_workflows_cli.py#L254) function in `argo_workflow_cli`.

**Review by cloudw — APPROVED** [review]

File: metaflow/plugins/kfp/kfp.py
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:411]: will handle in this PR https://github.com/zillow/metaflow/pull/240
with ticket [AIP-7535](https://zbrt.atl.zillow.net/browse/AIP-7535) WFSDK: Separate register and schedule workflow functions

**Pull request merged**
